### PR TITLE
Fix possible timeouts in `sh` tests with tsan

### LINKED ISSUE #68626: Stateless test `00653_verification_monotonic_data_load` is flaky
In my PR https://github.com/ClickHouse/ClickHouse/pull/68323 

2024-08-19 18:32:27 00653_verification_monotonic_data_load:                                 [ FAIL ] 600.05 sec.

The test times out in **ClickHouse Stateless tests (tsan, s3 storage)** 

The test is a shell script where the `clickhouse client` commands output is piped to a grep. Below is the server debug log for a command which was stuck for 4 minutes and then the test was killed due to hitting 600 seconds.

```
2024.08.19 18:28:11.330041 [ 10795 ] {} <Trace> TCPHandlerFactory: TCP Request. Address: [::1]:47212
2024.08.19 18:28:11.333131 [ 10795 ] {} <Debug> TCPHandler: Connected ClickHouse client version 24.9.0, revision: 54470, database: test_ltxmr5rm, user: default.
2024.08.19 18:28:11.333490 [ 10795 ] {} <Debug> TCP-Session: c36d5f04-e548-48ff-998d-f742b41219f6 Authenticating user 'default' from [::1]:47212
2024.08.19 18:28:11.333813 [ 10795 ] {} <Debug> TCP-Session: c36d5f04-e548-48ff-998d-f742b41219f6 Authenticated with global context as user 94309d50-4f52-5250-31bd-74fecac179db
2024.08.19 18:28:11.333966 [ 10795 ] {} <Debug> TCP-Session: c36d5f04-e548-48ff-998d-f742b41219f6 Creating session context with user_id: 94309d50-4f52-5250-31bd-74fecac179db
2024.08.19 18:28:11.405175 [ 10795 ] {} <Trace> TCP-Session: c36d5f04-e548-48ff-998d-f742b41219f6 Creating query context from session context, user_id: 94309d50-4f52-5250-31bd-74fecac179db, parent context user: default
2024.08.19 18:28:11.412546 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> executeQuery: (from [::1]:47212) (comment: 00653_verification_monotonic_data_load.sh) SELECT count() FROM signed_integer_test_table WHERE toUInt32(val) == 0; (stage: Complete)
2024.08.19 18:28:11.418827 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> Planner: Query to stage Complete
2024.08.19 18:28:11.420152 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> Planner: Query from stage FetchColumns to stage Complete
2024.08.19 18:28:11.423451 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> QueryPlanOptimizePrewhere: The min valid primary key position for moving to the tail of PREWHERE is 0
2024.08.19 18:28:11.424281 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Key condition: (toUInt32(column 0) in [0, 0])
2024.08.19 18:28:11.424598 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Filtering marks by primary and secondary keys
2024.08.19 18:28:11.425041 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Used generic exclusion search over index for part all_1_1_0 with 8 steps
2024.08.19 18:28:11.425290 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Selected 1/1 parts by partition key, 1 parts by primary key, 2/7 marks by primary key, 2 marks to read from 1 ranges
2024.08.19 18:28:11.425603 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Trace> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Spreading mark ranges among streams (default reading)
2024.08.19 18:28:11.426085 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Reading approx. 2 rows with 1 streams
2024.08.19 18:28:11.448629 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> executeQuery: Read 2 rows, 8.00 B in 0.039218 sec., 50.99699117752052 rows/sec., 203.99 B/sec.
2024.08.19 18:28:11.450063 [ 10795 ] {9c87b100-7c2a-404c-8cba-878a6e53e849} <Debug> TCPHandler: Processed in 0.047461623 sec.
2024.08.19 18:32:27.679498 [ 10795 ] {} <Debug> TCPHandler: Done processing connection.
```
At 18:32:27, the test was killed and hence the "Done processing connection" message is seen.  Notice the 4 minute gap. The previous command in the test passed successfully and it's debug output looks like :-

```
2024.08.19 18:28:10.386297 [ 10816 ] {5a70522e-5958-4ef3-9eb7-fcc7061c3193} <Debug> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Selected 1/1 parts by partition key, 1 parts by primary key, 2/7 marks by primary key, 2 marks to read from 1 ranges
2024.08.19 18:28:10.386602 [ 10816 ] {5a70522e-5958-4ef3-9eb7-fcc7061c3193} <Trace> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Spreading mark ranges among streams (default reading)
2024.08.19 18:28:10.386977 [ 10816 ] {5a70522e-5958-4ef3-9eb7-fcc7061c3193} <Debug> test_ltxmr5rm.signed_integer_test_table (66f2add6-d9c5-4eef-adf0-773f6c90e542) (SelectExecutor): Reading approx. 2 rows with 1 streams
2024.08.19 18:28:10.403503 [ 10816 ] {5a70522e-5958-4ef3-9eb7-fcc7061c3193} <Debug> executeQuery: Read 2 rows, 8.00 B in 0.02959 sec., 67.59040216289287 rows/sec., 270.36 B/sec.
2024.08.19 18:28:10.405171 [ 10816 ] {5a70522e-5958-4ef3-9eb7-fcc7061c3193} <Debug> TCPHandler: Processed in 0.036433273 sec.
2024.08.19 18:28:10.607434 [ 10816 ] {} <Debug> TCPHandler: Done processing connection.
2024.08.19 18:28:10.609280 [ 10816 ] {} <Debug> TCP-Session: ba83d9ed-0a90-4583-94c4-5e30def3029b Logout, user_id: 94309d50-4f52-5250-31bd-74fecac179db
```

The test's command looks like this :

`${CLICKHOUSE_CLIENT} --query="SELECT count() FROM signed_integer_test_table WHERE toUInt32(val) == 0;" 2>&1 |grep -q "2 marks to read from" && echo "monotonic int case: Int32 -> UInt32"
`

Maybe the last bytes of output in the TCP socket get stuck somewhere or something to do with socket close etc.

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Fix data race in clickhouse-client when it's piped to a process that terminated early

Close https://github.com/ClickHouse/ClickHouse/issues/68626

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 3d0614f2566e8616d3eb42895c33b1c514b375c1 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68791/3d0614f2566e8616d3eb42895c33b1c514b375c1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/b22423068b8d36b90168a9fd0712d852cc745500/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by vitlibar** [comment]
The first commit https://github.com/ClickHouse/ClickHouse/pull/68791/commits/3d0614f2566e8616d3eb42895c33b1c514b375c1 is enough. And the issue with failed tests can be solved by changing the lines
https://github.com/ClickHouse/ClickHouse/blob/27403207069f637a3653cd7554c52825d6e6e91f/programs/client/Client.cpp#L1194-L1196
to
```
        auto query_id = config().getString("query_id", "");
        if (!query_id.empty())
            client_context->setCurrentQueryId(query_id);
```

**Comment by vitlibar** [comment]
Is it possible to make a test reproducing that data race?

**Comment by vdimir** [comment]
> Is it possible to make a test reproducing that data race?

Actually, it’s reproducing in the CI, for example, in  [00653_verification_monotonic_data_load](https://play.clickhouse.com/play?user=play#CldJVEgKICAgIGZhbHNlIGFzIG9ubHlfbWFzdGVyClNFTEVDVCAqCkZST00gKApTRUxFQ1QKICAgIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHJudW0sCiAgICBzdWJzdHIoY29tbWl0X3NoYSwgMSw5KSBhcyBzaGEsCiAgICByb3VuZCh0ZXN0X2R1cmF0aW9uX21zIC8gMTAwMCwgMikgYXMgZHVyLAogICAgY2hlY2tfc3RhcnRfdGltZSwKICAgIGZvcm1hdFJlYWRhYmxlVGltZURlbHRhKGludERpdihub3coKSAtIGNoZWNrX3N0YXJ0X3RpbWUsIDM2MDApICogMzYwMCkgYXMgYWdvLAogICAgY2hlY2tfbmFtZSwKICAgIHRlc3RfbmFtZSwKICAgIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgMQogICAgQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPj0gbm93KCkgLSBJTlRFUlZBTCA3IERBWQogICAgQU5EIChwdWxsX3JlcXVlc3RfbnVtYmVyID0gMCBPUiBub3Qgb25seV9tYXN0ZXIpCiAgICBBTkQgdGVzdF9zdGF0dXMgIT0gJ1NLSVBQRUQnCiAgICBBTkQgdGVzdF9zdGF0dXMgTElLRSAnRiUnCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwopCldIRVJFIHRlc3RfbmFtZSBsaWtlICclMDA2NTNfdmVyaWZpY2F0aW9uX21vbm90b25pY19kYXRhX2xvYWQlJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIERFU0MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==)

The problem is that TSAN reports for the client are not collected when a bug is reproduced. Instead, it is often only detected as a test timeout in most cases (so, there are probably other tests that time out for the same reason). So, we need to address the issue with sanitizer reports for the client in the CI in general.

File: programs/client/Client.cpp [hunk 1]
```diff
@@ -348,6 +348,7 @@ try
         // which is required for client timezone (pushed from server) to work.
         auto thread_group = std::make_shared<ThreadGroup>();
         const_cast<ContextWeakPtr&>(thread_group->global_context) = global_context;
+        const_cast<ContextWeakPtr&>(thread_group->query_context) = client_context;
```
Comment (vdimir) [inline programs/client/Client.cpp:351]:
```suggestion
        const_cast<ContextWeakPtr&>(thread_group->query_context) = client_context;
        const_cast<ContextWeakPtr&>(thread_group->session_context) = client_context;
```

**Comment by vdimir** [comment]
In the meantime, let's try to fix race bu just properly synchronizing formatting and main threads, https://github.com/ClickHouse/ClickHouse/pull/69186

**Pull request closed without merging**